### PR TITLE
Update `W605` to check in f-strings

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR updates the `W605` (invalid-escape-sequence) to check inside f-strings. It also adds support to report violation on invalid escape sequence within f-strings w.r.t. the curly braces. So, the following cases will be identified:

```python
f"\{1}"
f"\{{1}}"
f"{1:\}"
```

The new CPython parser also gives out a syntax warning for such cases:

```
fstring.py:1: SyntaxWarning: invalid escape sequence '\{'
  f"\{1}"
fstring.py:2: SyntaxWarning: invalid escape sequence '\{'
  f"\{{1}}"
fstring.py:3: SyntaxWarning: invalid escape sequence '\}'
  f"{1:\}"
```

Nested f-strings are supported here, so the generated fix is aware of that and will create an edit for the proper f-string.

## Test Plan

Add new test cases for f-strings.

fixes: #7295

### CODE REVIEW

**Comment by dhruvmanila** [comment]
Current dependencies on/for this PR:
* main
  * **PR #7035** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7035" target="_blank">Graphite</a> 
    * **PR #7013** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7013" target="_blank">Graphite</a> 
      * **PR #6659** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/6659" target="_blank">Graphite</a> 
        * **PR #7041** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7041" target="_blank">Graphite</a> 
          * **PR #7211** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7211" target="_blank">Graphite</a> 
            * **PR #7263** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7263" target="_blank">Graphite</a> 
              * **PR #7325** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7325" target="_blank">Graphite</a> 
                * **PR #7326** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7326" target="_blank">Graphite</a> 
                  * **PR #7327** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7327" target="_blank">Graphite</a> 
                    * **PR #7328** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7328" target="_blank">Graphite</a> 
                      * **PR #7329** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7329" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/7329?utm_source=stack-comment).

File: crates/ruff/src/rules/pycodestyle/rules/invalid_escape_sequence.rs [hunk 1]
```diff
@@ -77,9 +95,35 @@ pub(crate) fn invalid_escape_sequence(
...
+                // ```python
+                // f"foo\"
+                // ```
+                //
+                // Here, there won't be any `FStringMiddle` because it's an unterminated
+                // f-string.
+                let Some(next_char) = locator.after(tok_range.end()).chars().next() else {
+                    continue;
+                };
+                next_char
```
Comment (MichaReiser) [inline crates/ruff/src/rules/pycodestyle/rules/invalid_escape_sequence.rs:123]: Nit: The `let` else is a bit more complicated to read because I first read the exceptional case before the "normal" flow

What we want to express here is, take the `next_char` if any and otherwise return

```rust
if let Some(next_char) locator.after(tok_range.end()).chars().next() {
    next_char
} else {
    continue;
}
```

File: crates/ruff/src/rules/pycodestyle/rules/invalid_escape_sequence.rs [hunk 2]
```diff
@@ -77,9 +95,35 @@ pub(crate) fn invalid_escape_sequence(
...
+                //
+                // Now, if we're at the end of the f-string itself, the lexer wouldn't
+                // have emitted the `FStringMiddle` token in the first place. For example,
+                //
+                // ```python
+                // f"foo\"
+                // ```
+                //
+                // Here, there won't be any `FStringMiddle` because it's an unterminated
+                // f-string.
```
Comment (MichaReiser) [inline crates/ruff/src/rules/pycodestyle/rules/invalid_escape_sequence.rs:119]: I'm a bit confused by this comment. It says that there won't be any `FStringMiddle` token but this branch handles f-string middle tokens. How is this related?

File: crates/ruff/resources/test/fixtures/pycodestyle/W605_2.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+regex = f'\\\_'
+value = f'\{{1}}'
+value = f'\{1}'
+value = f'{1:\}'
+
+# Okay
+value = rf'\{{1}}'
+value = rf'\{1}'
+value = rf'{1:\}'
```
Comment (MichaReiser) [inline crates/ruff/resources/test/fixtures/pycodestyle/W605_2.py:51]: Can we add some tests for nested fstrings?

File: crates/ruff/src/rules/pycodestyle/rules/invalid_escape_sequence.rs (hunk 2 above)
Comment (dhruvmanila) [reply crates/ruff/src/rules/pycodestyle/rules/invalid_escape_sequence.rs:119]: That is for the code snippet mentioned above. I'm making a case to prove that the next character is always going to be inside the f-string as a whole. The code snippet is:

```python
f"foo\"
```

Here, the `FStringMiddle` token won't be emitted because it's an unterminated string literal as the end quote is being escaped. So, the character coming after a `\` is always going to be part of the f-string.

**Review by MichaReiser — APPROVED** [review]
